File: ABC049C.hs
{- ABC049C - Daydream -}
import qualified Data.ByteString.Char8 as B8
import           Prelude               hiding (words)

words :: [B8.ByteString]
words = map B8.pack [ "dream", "dreamer", "erase", "eraser" ]

solver :: B8.ByteString -> B8.ByteString
solver s =
  if go 0 then B8.pack "YES" else B8.pack "NO"
  where
    go idx | idx == B8.length s = True
    go idx =
      let
        s' = B8.drop idx s
        cond w = B8.isPrefixOf w s'
        nextIdx w = idx + B8.length w
      in
        any (== True) [if cond w then go (nextIdx w) else False | w <- words]

main :: IO ()
main = B8.putStrLn . solver =<< B8.getLine

File: ABC049C.cpp
// ABC049C - Daydream
#include <iostream>
#include <string>
#include <string_view>

bool walk(std::string_view s, std::size_t offset)
{
    static std::string_view words[] = {"dream", "dreamer", "erase", "eraser"};

    if (s.length() == offset)
        return true;
    else if (s.length() < offset)
        return false;
    else
    {
        for (auto w : words)
        {
            const auto len = w.length();
            if (w.compare(0, len, s, offset, len) == 0)
            {
                if (walk(s, offset + len))
                {
                    return true;
                }
            }
        }
        return false;
    }
}

bool solver(std::string_view s)
{
    return walk(s, 0);
}

int main()
{
    std::string s;
    std::cin >> s;
    std::cout << (solver(s) ? "YES" : "NO") << std::endl;
    return 0;
}
